[#5] 가게 추가, 수정, 삭제 구현

seohoon-dev opened the pull request:
#5

seohoon-dev committed:
가게 추가, 수정, 삭제 구현

seohoon-dev committed:
유저 내용 수정

- 유저 관련 메소드명 수정
- Exception 내용 수정

seohoon-dev committed:
세션 내용 추가

- 가게 정보 등록 및 변경 시 로그인 필요
- 본인 가게 정보만 변경 가능

seohoon-dev committed:
비밀번호 암호화

- sha256

f-lab-dev commented on src/main/resources/mapper/RestaurantMapper.xml:
```diff
@@ -0,0 +1,48 @@
+<mapper namespace="com.hoon.foodrocket.mapper.RestaurantMapper">
+
+    <!-- 가게 목록 -->
+    <select id="findAll" resultType="com.hoon.foodrocket.domain.Restaurant">
```
만약 데이터가 1억건이면 이 쿼리는 1억건을 모두 조회하게 됩니다.
그러면 데이터베이스 엄청난 부하가 가해지고 서비스에 장애가 발생합니다.

우리가 어플리케이션을 사용할때 존재하는 모든 가게를 불러오지는 않습니다.
아래의 필터링을 거치게 되겠네요
1. 지역
2. 카테고리 (중식, 한식 등)
3. 가게 오픈여부

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/RestaurantService.java:
```diff
@@ -0,0 +1,79 @@
+    private RestaurantMapper restaurantMapper;
+
+    @Transactional
+    public List<Restaurant> getRestaurants() {
```
이 기능이 꼭 필요한지 생각해보시면서 개발해보시는게 좋습니다~
최고관리자라고 하더라도 전국에 등록된 모든 가게를 한 화면에서 조회하지는 않을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/RestaurantService.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Transactional
+    public void addRestaurant(String name, String address, String ownerEmail) {
```
이 메소드의 `비즈니스 로직`이 무엇인지 파악하시고 로직을 세우시는게 좋을 것 같네요.
전체적으로 이 클래스의 메소드들은 단순히 dao를 호출하기만 하고 비즈니스 로직이 없습니다.

이 메소드를 예로 들자면 사장이 가게를 등록하려고 하면
1. 기존에 사장에 동일한 주소에 동일한 가게를 중복으로 올리는 것인지 검사해야할 것 같습니다.
2. 사장이 등록할 수 있는 가게의 최대 개수를 제한해야할텐데 그것도 검사해주면 좋을 것 같습니다.
3. 모든 요건을 충족했을 때 가게를 등록시킵니다

f-lab-dev replied:
추가적으로 가게의 정보에는 더 많은게 들어갈 것 같은데 어떤 정보가 더 필요할지 고민해보시고, `DTO`를 사용하시는게 좋을 것 같습니다.

f-lab-dev replied:
또한 이 클래스에서 전체적으로 메소드를 작명할 때 `Service` 레이어는 비즈니스로직을 담는 역할의 클래스이니 DB용어보다는 비즈니스 용어에 가깝게 이름을 지어주는 것이 좋습니다.

`add`보다는 `register` 등의 용어가 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/RestaurantController.java:
```diff
@@ -0,0 +1,68 @@
+    public ResponseEntity<?> create(@RequestBody Restaurant resource, HttpSession session) {
+        String loginUserEmail = HttpSessionUtil.getLoginUserEmail(session);
+
+        if (loginUserEmail == null) {
```
모든 유저가 다 가게를 등록할 수 있지는 않을 것 같습니다~ 사장회원을 구분해주면 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/domain/Restaurant.java:
```diff
@@ -0,0 +1,34 @@
+    @NotEmpty
+    private String ownerEmail;
+
+    public boolean matchOwnerEmail(String loginUserEmail) {
```
보통 boolean이 리턴값일 때에는 앞에 `is`를 붙여줍니다.
`isMatchOwnerEmail`이 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/domain/Restaurant.java:
```diff
@@ -0,0 +1,34 @@
+
+@Builder
+@Getter
+@NoArgsConstructor
```
`Setter`가 없으므로 `NoArgsConstructor`도 없어도 될 것 같습니다~

seohoon-dev replied:
Setter와 NoArgsConstructor 의 상관관계를 잘 모르겠습니다.
NoArgsConstructor 는 매개변수가 없는 생성자를 만들어줄때 필요한 거 아닌가요?

f-lab-dev replied:
매개변수가 없는 생성자를 쓴다는 것은 내부의 값이 비어있다는 뜻입니다.
그럼 내부에 값을 세팅해주려면 `setter`를 사용해야겠죠. 그래서 보통 `immutable` 클래스를 만들때는 getter와 AllArgConstructor를 사용합니다.

seohoon-dev committed:
사장, 회원 구분

- 사장 관련 코드 분리

seohoon-dev committed:
가게 관련 코드 수정

- 가게 최대 등록 수 추가
- 가게 정보 추가
- 메소드명 변경

seohoon-dev committed:
지역별 가게 목록

- region 조건을 통해 가게 목록 필터링

seohoon-dev committed:
카테고리별 가게 목록

- category 조건을 통해 가게 목록 필터링

seohoon-dev committed:
리뷰 내용 반영

- throw를 통해 예외 발생

seohoon-dev committed:
Merge branch 'issue/5' of https://github.com/f-lab-edu/food-rocket into issue/6

# Conflicts:
#	src/main/java/com/hoon/foodrocket/application/UserService.java

seohoon-dev commented on src/main/resources/mapper/RestaurantMapper.xml:
```diff
@@ -6,14 +6,15 @@
+    <select id="getRestaurantsByAddress" parameterType="String" resultType="com.hoon.foodrocket.domain.Restaurant">
+        select id, name, address, owner_email ownerEmail, information, phone_number phoneNumber, open_time openTime, close_time closeTime
         from restaurant
+        where address like concat('%',#{region},'%')
```
주소의 형식을 어떻게 표현해야될지 고민입니다.
현재는 address에 '서울시 용산구 ~~' 와 같은 형식으로 들어가있고, 쿼리 조건으로 region이 address에 포함되면 목록을 보여주는 식입니다.
ex) address='서울시 용산구 ~~', region='용산'

자바의 contains와 비슷한 방식으로 만들었는데 성능적으로 문제가 되는 부분이 있을까요?

f-lab-dev replied:
네 인덱스가 잡혀있지 않은 address 전체에 like 연산을 하게되면 꽤 많은 부하를 주게 됩니다.
원래 address는 본인이 입력하는게 아니라 나라에서 제공하는 주소 DB가 있습니다. 이 DB를 이용해 적절하게 데이터를 가지고 사용하시면 됩니다.
`zipcode`를 키워드로 찾아보시면 될 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+    @Autowired
+    OwnerMapper ownerMapper;
+
+    @Transactional
```
이 어노테이션은 면접에도 많이나오고 아주 중요한 어노테이션입니다. 하지만 달아주기만 하면 마법같이 편하게 트랜잭션을 사용할 수 있어서 많은 사람들이 이 어노테이션에 대해 깊게 알지 못한채 쓰고있습니다. 면접에서도 그래서 대답하지 못하는 경우가 많고요.

이 어노테이션의 위에 이 어노테이션이 어떤 역할을 하는지 공부해보시고 역할을 주석으로 달아주시면 좋을것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
```
이 어노테이션도 어떤 역할을 하는건지 주석으로 달아주시면 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+
+@Service
+public class OwnerService {
+    @Autowired
```
`@Autowired`도 있고, `@Resource`도 있습니다. 이 어노테이션의 차이도 공부해보시면 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+            throw new IllegalStateException("이미 등록된 정보(사장)입니다.");
+        }
+
+        Owner builder = Owner.builder()
```
이펙티브자바를 보면 빌더패턴을 사용하는 이유가 나옵니다. 그 것도 보시고 빌더패턴을 사용했을때 뭐가 좋은지 스스로도 생각해보시고 느껴보시면 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OwnerController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<?> create(@RequestBody Owner resource) {
```
기왕이면 제네릭 타입에 명확한 타입이 들어있으면 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OwnerController.java:
```diff
@@ -0,0 +1,58 @@
+
+        ownerService.registerOwner(email, name, password);
+
+        return ResponseEntity.status(HttpStatus.CREATED).build();
```
이런 경우는 `HttpStatus.CREATED`만 리턴하는 것도 같은 결과를 내보냅니다

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        if (!owner.isMatchPassword(password)) {
+            throw new IllegalArgumentException("비밀번호가 일치하지 않습니다.");
```
좋은 예외 선택입니다

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -0,0 +1,55 @@
+            throw new IllegalStateException("정보(사장)가 없습니다.");
+        }
+
+        if (!owner.isMatchPassword(password)) {
```
이런 경우는 가독성을 위해 owner에다가 `isNotMatchPassword` 메소드를 구현해주는 트릭도 있습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/mapper/OwnerMapper.java:
```diff
@@ -0,0 +1,17 @@
+    Owner getOwnerFromEmail(String email);
+
+    // 회원가입
+    void registerOwner(Owner owner);
```
Mapper는 비즈니스로직보단 DB에 더 가까운 클래스이므로 단어선택도 DB에 가까운 용어를 써주는게 좋습니다~
Mapper, Service, Controller 의 `역할`과 `책임`에 대해 찾아보시고 고민해보시면 좋을 것 같습니다
`insertOwner`등의 이름이 적절해보이네요

seohoon-dev committed:
Merge pull request #12 from f-lab-edu/issue/6

[#6] 비밀번호 암호화

seohoon-dev committed:
리뷰 내용 반영

- aop 적용
- 메소드명 변경
- 어노테이션 주석 추가

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * "Aspect"
+ * Aspect 역할을 명시한다.
+ * 핵심 기능에서 부가 기능을 분리하기 위한 개념이다.
```
적용은 잘 하셨지만 `@Aspect`를 달아주면 내부적으로 어떤일이 일어나는지 상세하게 적어주시면 좋을 것 같습니다~
이해하고 사용하는게 제일 중요해서요

seohoon-dev replied:
Aspect 내부에 정의된 Target이나 Retention은 확인할 수 있었습니다. aspectj 문서도 봤지만 동일한 내용만 볼 수 있었는데 참고할만한 문서가 있을까요?

f-lab-dev replied:
`핵심 기능에서 부가 기능을 분리하기 위한 개념이다.` 라기보다는 AOP를 좀 더 공부해보면 `프록시`라는 단어가 나옵니다. 이 `프록시`가 어떻게 입혀지는지 토비의 스프링쪽을 조금 더 참고해보시면 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -0,0 +1,38 @@
+ * Aspect 역할을 명시한다.
+ * 핵심 기능에서 부가 기능을 분리하기 위한 개념이다.
+ *
+ * "Component"
```
`@Aspect`와 `@Component`를 왜 같이 써야할까요?

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OwnerService.java:
```diff
@@ -2,12 +2,26 @@
+/**
+ * "Service"
+ * 해당 클래스가 Service 클래스임을 명시한다.
+ * 비지니스 로직이 들어가는 Service 빈을 등록한다.
```
어떤 단계에서 빈이 등록되는지, 기본적으로 빈 이름은 어떻게 정해지는지도 설명해주시면 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/CheckOwnerLogin.java:
```diff
@@ -0,0 +1,4 @@
+package com.hoon.foodrocket.aop;
+
+public @interface CheckOwnerLogin {
```
이렇게 하면 가게 사장일때만 검증할 수 있게됩니다.
일반 회원을 검증하는 기능도 같이 사용할 수 있게하면 좋을 것 같습니다.

seohoon-dev replied:
CheckUserLogin 와 같은 인터페이스를 추가하면 좋겠다는 의미인가요?

f-lab-dev replied:
인터페이스 추가라기보다는 이 어노테이션을 더 범용적으로 만들면 좋겠다는 말이였습니다

seohoon-dev committed:
리뷰 내용 반영

- 로그인 검증 어노테이션 수정
- 주소 DB 적용
- 유저 vo에 address, region 추가
- 유저의 지역과 음식 카테고리를 조건으로 가게 목록을 보여줌

seohoon-dev committed:
[#13] 메뉴

- 메뉴 등록 및 수정, 삭제
- 가게 상세정보에서 메뉴 목록 확인

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -24,15 +26,24 @@
         HttpSession session = ((ServletRequestAttributes) RequestContextHolder.currentRequestAttributes()).getRequest().getSession();
 
-        String loginOwnerEmail = HttpSessionUtil.getLoginOwnerEmail(session);
+        if (loginType.type().equals("owner")) {
```
만약 권한이 100개가 된다면 if문도 100개가 될텐데 분기를 태우는것보다는 여기에 다형성을 이용해보면 어떨까요?
`LoginVerificationService` 같은 인터페이스를 만들어서 권한별로 구현하면 개구리책에 나오는 객체지향의 원칙을 그대로 맞춰줄 수 있습니다.

seohoon-dev replied:
LoginVerificationService 인터페이스에 ownerLoginCheck(), userLoginCheck() 와 같은 메서드를 정의하고, LoginCheckAspect 클래스에서 오버라이딩해서 사용하라는 뜻인가요?

제가 이해한 방식이면 결국 if문 사용은 동일한거 같은데 잘못 이해했다면 말씀해주세요.
```
public class LoginCheckAspect implements LoginVerificationService {
// ...
    public void loginCheck(LoginType loginType) {
        // ...
        if (loginType.type().equals(Type.OWNER)) {
            ownerLoginCheck(session);
        } else if (loginType.type().equals(Type.USER)) {
            userLoginCheck(session);
        }
    }

    @Override
    public void ownerLoginCheck(HttpSession session){...}

    @Override
    public void userLoginCheck(HttpSession session) {...}
}
```

f-lab-dev replied:
음 제 설명은 `HttpSessionUtil.getLoginOwnerEmail(session);` 처럼 레벨의 개수 n개만큼 `HttpSessionUtil`의 메소드가 늘어나야하는 것을 우려한 말이였습니다. 

어차피 비슷한 역할을 하고있으니 if문으로 분기처리하지 말고 메소드를 1개로 통일시켜줄 수 있을지, 안됀다면 다형성을 이용해서라도 분기를 최대한 줄이려는거였습니다. 

핵심만 말하자면 이 if문을 없앨 수 있는 방법은 무엇무엇이 있을까요?

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/RegionService.java:
```diff
@@ -19,9 +19,8 @@
     }
 
     @Transactional
```
여기는 insert 한 번만 일어나는데 `@Transactional`을 달아줄 필요가 있을까요?

f-lab-dev commented on src/main/resources/mapper/RegionMapper.xml:
```diff
@@ -12,13 +12,18 @@
+            restaurant_id
         )
+        values
+        <foreach item="item" index="index" collection="list" separator=",">
```
벌크 insert를 사용한게 좋네요. 이 부분도 이력서 정리할때 insert 성능 향상으로 같이 정리해주면 좋을 것 같습니다.

f-lab-dev commented on src/main/resources/mapper/RestaurantMapper.xml:
```diff
@@ -9,7 +9,11 @@
+        where id in (
+            select restaurant_id
+            from region
+            where name = #{region}
```
여기서 체크해봐야할 점은 아래의 쿼리가 효율적이냐를 봐야합니다. 아래의 두개를 체크해보면 좋을 것 같네요
1. name에 인덱스가 걸려있는가
2. 결과로 리턴될 레스토랑은 몇개나 될 것인가

```
select restaurant_id
from region
where name = #{region}
```

seohoon-dev replied:
name에는 인덱스를 걸어둔 상태이고, 리턴될 레스토랑 갯수는 지역별로 가져오지만 한 지역에 1만 개의 가게가 있으면 한꺼번에 가져올 것 같습니다. 페이징 처리가 필요할까요?

f-lab-dev replied:
네 더 크게 생각하면 1억개가 있으면 1억개를 모두 가져오겠군요.
페이징같은 이거에 대한 최적화를 고민해보시면 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginType.java:
```diff
@@ -0,0 +1,12 @@
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface LoginType {
+    String type();
```
이 type을 enum으로 만들면 어떨까요? 그럼 더 소스가 명확해질 것 같습니다.
String으로 쓰게되면 오타가 있어도 컴파일타임에 잡지 못하게 되어서 서비스가 동작하고 있을때 에러가 발생할 확률이 높기 때문입니다

seohoon-dev committed:
리뷰 내용 반영

- LoginVerificationService 인터페이스 생성
- 불필요한 @Transactional 제거
- LoginType에 type을 enum으로 변경

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -21,7 +21,7 @@
 @Aspect
 @Component
-public class LoginCheckAspect {
+public class LoginCheckAspect implements LoginVerificationService {
```
여기에 인터페이스를 입히지는 않아도 됩니다. 이거는 금요일에 한번 설명드리는게 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -30,20 +30,29 @@
 
-        if (loginType.type().equals("owner")) {
-            String loginOwnerEmail = HttpSessionUtil.getLoginOwnerEmail(session);
+        if (loginType.type().equals(Type.OWNER)) {
```
일단 enum을 사용한것은 깔끔하네요~ enum은 `==`로 비교하셔도 됩니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/Type.java:
```diff
@@ -0,0 +1,5 @@
+package com.hoon.foodrocket.aop;
+
+public enum Type {
```
그냥 `Type`이라고 이름을 지어두면 나중에 이 코드를 유지보수하는 다른 개발자는 의미를 헷갈려할 수 있습니다.
그래서 클래스의 목적에 맞게 이름을 잘 지어줘야하는데요.
이 enum의 목적은 `유저의 권한레벨`을 나타내는거니까 이 목적에 맞는 이름을 지어보시는게 좋을 것 같습니다

seohoon-dev committed:
리뷰 내용 반영

- 쿼리를 호출하는 반복문 최소화
- 불필요한 Transactional 제거

seohoon-dev committed:
Merge branch 'issue/5' of https://github.com/f-lab-edu/food-rocket into issue/13

seohoon-dev committed:
메뉴 쿼리 변경

- 기존 insert와 update를 하나로 합침
- delete를 bulk로 변경

seohoon-dev committed:
Merge pull request #16 from f-lab-edu/issue/13

[#13] 메뉴

seohoon-dev committed:
리뷰 내용 반영

- 다형성을 통해 유저 권한 별 로그인 검증 절차를 만듬
- 가게 목록 호출 시 커서 기반 페이징을 적용
- enum Type을 UserAuthorityLevel로 변경

f-lab-dev commented on src/main/resources/mapper/MenuMapper.xml:
```diff
@@ -12,38 +12,33 @@
             #{item.restaurantId}
             )
         </foreach>
+        on duplicate key update
```
이런것도 이력서에 쓸만한 튜닝내용입니다~ 잘 정리해보시고 만났을때 같이 다듬어보면 좋겠네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -30,29 +41,21 @@
-        }
+        map.get(loginType.level()).loginVerification(session);
 
+        System.out.println("map: " + map);
```
요거는 로거를 사용하시는게 좋습니다. 앞으로 `System.out.println`은 아예 사용을 안한다는 생각으로 코딩하시는게 좋습니다~
자바 성능 튜닝 이야기에 관련 내용이 나옵니다

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -30,29 +41,21 @@
-        } else if (loginType.type().equals(Type.USER)) {
-            userLoginCheck(session);
-        }
+        map.get(loginType.level()).loginVerification(session);
```
이제 다형성의 장점을 체험해보셨으니 이거를 "어떻게 더 잘할까"를 생각해보셔야하는데요.
map을 사용하면 없는 `level`을 넣으면 null이 나오기 때문에 에러가 발생할 가능성이 있습니다.
그럼 이 것을 어떻게 주입하는 것이 좋을까요?
답은 정해져있지 않습니다. 다만 고민해보시면 좋을 것 같아요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/aop/LoginCheckAspect.java:
```diff
@@ -30,29 +41,21 @@
-        }
-    }
+@Configuration
+class CollectionConfig {
```
같은 클래스안에 두는건 좋지 않습니다.
따로 Config 클래스를 빼는게 좋을 것 같네요~

seohoon-dev committed:
리뷰 내용 반영

- System.out.println 제거(로거를 사용할 수 있도록 할 것)
- config 클래스 분리

f-lab-dev reviewed (approved):
config를 분리는 했지만 `왜` 분리해야 좋은지는 얘기를 안했었는데요.
왜 좋을지도 유지보수성 측면에서 고민해보시면 좋을 것 같습니다
